Quick Vent, LED updates

nomakewan opened the pull request:
This PR does the following:

-Adds placeholder sounds for Slime Tether, Shock Blast, and Meson Collider
-Adds ability for standalone wand to play additional Proton Stream impact SFX
-Adds new unique Quick Vent function which is shorter than overheating fully
-Adds sanity check to EEPROM read function that will reset the EEPROM to default if CRC does not match
-Changes LED calls which were only on/off from analogWrite to digitalWrite
-Cleans up 1984/1989 functionality on Hat Light 1
-Cleans up Proton Pack initialization process
-Corrects default EEPROM values for various wand settings
-Corrects EEPROM LED function name mismatch
-Corrects Hat Light 2 functionality and adds 1984/1989 functionality
-Corrects LED documentation in Neutrona Wand header file
-Corrects misplaced steam SFX and N-Filter strobe when overheating in Slime Mode
-Corrects Power Cell functionality on first Proton Pack boot-up
-Corrects Power Cell functionality when hot-swapping wands
-Corrects Power Cell timing when settings are loaded from EEPROM
-Corrects switching firing modes or exiting menus causing unexpected Power Cell illumination
-Corrects wand-side Proton Stream additional SFX not obeying user setting
-Removes random impact effects from CTS firing
-Switches several timer start calls to restart for efficiency

nomakewan commented:
Latest push adds a sanity check to the function that reads from the EEPROM so that if the CRC does not match, the EEPROM will be cleared to default values.

The first time I loaded this firmware to my wand, the volume setting was incorrect due to a shift in where the address for the volume was stored. I resolved it already by just entering the EEPROM config menu and clearing the EEPROM, but hopefully by including this new sanity check the requirement to manually clear the EEPROM will be alleviated.

gpstar81 reviewed (approved):
In the quick vent, the cyclotron lights turn off for that sequence, I'm not sure if that is intended? Otherwise if not, I can track it down afterwards. There is a powercell bug that was introduced awhile back in the audio branch where sometimes it refuses to light up, sometimes it does. Something that I will track down afterwards.

Anyhow, I see the updated quick vent will be much appreciated by cos-players.

nomakewan commented on source/NeutronaWand/NeutronaWand.ino:
```diff
@@ -782,14 +811,20 @@ void startVentSequence() {
+  }
 
-  delay(100); // Really should avoid this if possible since it will corrupt our ezButton states
+  delay(100); // TODO: Really should avoid this if possible since it will corrupt our ezButton states
```
I haven't tried with it removed, and you're right, I'm not sure why it's there either. I didn't put it into Quick Vent and that seems to work fine, so I'll just remove it.

Speaking of not knowing why things are there, why do we delay 50ms before playing firing sounds and 100ms before playing end-firing sounds? It doesn't affect much except that if you tap Intensify and release it faster than 50ms, no firing sound will play but the end-firing sound will play. Not a major thing, but just wondered why it's there.

gpstar81 replied:
I think that old delay was for some sound timing and I was trying to get things out quickly.

The ms timer delay for firing sounds is for helping the serial communication buffer to the audio board from not overflowing (or it was the original intention). If I recall some sounds play instantly while a few more get mixed in slightly after. There was a issue with the wav trigger track reporting as it always sends data back which was causing a major crash bug in September 2023. It was later fixed by fixing some serial data checks from incoming data. It shouldn’t be a issue for GPStar audio. Though it may be ok to remove those delays or make them super short, will need to be tested if it is decided to remove them though.

nomakewan replied:
Sounds like something to be left for after 5.0.0, then. Let's not break too much before a major release, hahaha. Thanks!

Looking into the cyclotron thing now; no, it's not intended behavior. I didn't think I'd touched the cyclotron except to call the function that reverts it to normal speed, so I'm not sure why it shuts off. Poking around.

nomakewan commented:
I moved the venting sequence out of the cyclotron control loop; hopefully that resolves the cyclotron going dark while venting as that was not intended behavior. Essentially while quick venting the pack and wand should remain fully functional but unable to fire again until the sequence completes, and the cyclotron speeds should revert to normal. At the moment this doesn't do anything because you'd have to stop firing to use Quick Vent (which already reverts the cyclotron speed to normal), but once there's a heat management system it will become more relevant.

nomakewan commented:
Okay, I believe I've fixed the powercell issue now. packOffReset() was normally resetting the timer for the powercell 100% of the time because it was being called in an endless loop whenever the pack was off. This was corrected in the audio branch so that it's only called once, but the problem was that it was not called the very first time the Talentcell was powered on because the header file defined b_reset_start_led as true.

So I just took the missing timers from packOffReset() and added them to the pack's initialization process. Hopefully that resolves the issue.

nomakewan commented:
Yep, was just another missed call. Should be okay now.

gpstar81 commented:
> Yep, was just another missed call. Should be okay now.

Working good.

gpstar81 commented:
actually did some more testing, it's still there.

nomakewan commented:
That's odd; it should be setting all the LEDs to black before restarting the timer. Is it only happening when the Talentcell is first turned on, or have you found it happening at any other times? Like say, does it happen when you overheat due to firing and everything ramps down and shuts off before turning back on again?

gpstar81 commented:
When there is no wand, it seems to come on then off after two seconds.

Then when I connect a wand, it comes back on during the sync.

Though sometimes it doesn't come on during a power startup.

I wonder if it's something to do with the EEPROM loading and resetting something and something might be called at the same time while the fast led update is skipped leaving it on?

nomakewan commented:
I'm gonna rest on it a bit but I think you're onto something. There is indeed a call to reset some LEDs after reading the EEPROM so that could be part of it. I'm going to convert this PR to a draft for the moment until it can be resolved. If you get to it before me though, feel free to commit!

nomakewan commented:
Okay, as a note I found the problem, and it is indeed part of the serial data handshake procedure. However, the reason why that is there is because at the moment there's no failsafe for yanking a wand while in the settings menu. I'll look into implementing that failsafe functionality and fix the handshake in a few hours once I'm back.

gpstar81 commented:
hmm, I thought the pack would return back to its previous mode or proton mode while in the setting menu and the wand get's disconnected?

nomakewan commented:
Okay. After staring at how each of the four menus (Settings while on, Settings while off, EEPROM LED, EEPROM Config) work as well as how the various handshakes and reset vectors work, I believe the changes I've just pushed should put this one to bed once and for all. At least until it's time to clean up some more code, perhaps.

I've left commented out the ability for the pack to save any in-progress settings it had in an EEPROM menu if it was in the menu when a wand was hot-swapped. At the moment the pack actually has no clue whether it's in the EEPROM Config menu or not as that entire menu is handled by the wand, however, so to do this correctly a new variable would have to be introduced to let the pack keep track of whether it's in the menu or not. So like I said, for now, it's just commented out.

Please give this a test and see if it works as expected. Especially test:

1. Initial Talentcell power-on w/o wand.
2. Initial Talentcell power-on w/ wand.
3. After turning on Talentcell w/o wand, plug in wand.
4. Hotswap wand while pack is off.
5. Hotswap wand while pack is on.
6. Hotswap wand while in settings menu with wand and pack off.
7. Hotswap wand while in settings menu with wand and pack on.
8. Hotswap wand while in EEPROM LED Menu.
9. Hotswap wand while in EEPROM Config Menu.

gpstar81 commented:
#1 through #9 all works correctly.

Just one odd thing I notice. When doing the first turn on of the pack (ion arm switch), the powercell automatically lights up the first and 2nd LED together at the same time before doing the regular ramping up. When shutting the pack down and turning back on (not talent cell, but the red ion arm switch), it is correct with the correct pause between the 1st and 2nd led during startup.

The rest seems good.

nomakewan commented:
> #1 through #9 all works correctly.
> 
> Just one odd thing I notice. When doing the first turn on of the pack (ion arm switch), the powercell automatically lights up the first and 2nd LED together at the same time before doing the regular ramping up. When shutting the pack down and turning back on (not talent cell, but the red ion arm switch), it is correct with the correct pause between the 1st and 2nd led during startup.
> 
> The rest seems good.

Does that happen every single time you start the pack from the ion arm switch, or only in certain pack/wand configurations?

gpstar81 commented:
Every time

nomakewan commented:
New update rearranges the pack initialization process and fixes two checks elsewhere in the code that were looking at incorrect variables. This significantly cleans things up and make it a lot easier to follow what exactly is happening and why, but I am not confident that this cleanup alone will resolve that odd Power Cell behavior on first ion arm switch powerup. In fact, if the bug is actually somewhere in the reset function it could expand this bug to happen on every ion arm startup instead.
